1021 - Add multiple choice answer type for Question node type

harry02031 opened the pull request:
## Changes
This PR allows the author to select a multiple choice answer type to the question node type (in the form of checkboxes if they want the user to select more than one answer or in the form of radio buttons if they only want the user to select one answer). The author is able to type in the values of the choices, change the ordering of the choices and also upload images for each choice if they choose. The user is able to then click on the lightbox and select answer to the question in the form of checkboxes or radios depending on whatever the author selects during the authoring process.

* Authoring: the author is able to able to store all their input information in node.typeData.options.multipleChoice
<img width="565" alt="Screen Shot 2021-05-26 at 10 48 35 AM" src="https://user-images.githubusercontent.com/30706624/119707793-677e0b80-be10-11eb-9824-61b7c721c16d.png">

* User: the answer selected by the user is saved in quiz[0].entries as multipleChoiceId: { [multipleChoiceId]: "response" } to make it similar to how text and audio responses are stored in the depended PR 991
<img width="563" alt="Screen Shot 2021-05-26 at 10 49 21 AM" src="https://user-images.githubusercontent.com/30706624/119708760-5255ac80-be11-11eb-84e1-12bd9ad030d9.png">


## Screenshot(Video of Feature)

https://user-images.githubusercontent.com/30706624/119707676-461d1f80-be10-11eb-8502-ed1614b3a136.mov


## Issue Linkage
Closes #1021
## PR Dependency
Depends on #994
## Automated Testing
* should be able to complete with a radio answer
* should be able to complete with a checkbox answer
* should be able to complete with a checkbox answer with images
* should be able to complete with a radio answer with images

Frozemint commented:
Is the "Show Answer to a Previous Activity First" feature supported yet? I can't seem to be able to get it working.

Also, currently it's possible to create a multiple choice question with no options. Is this intended behaviour?

harry02031 commented:
Hey @Frozemint, you can git pull to get the latest version that addresses the issues you mentioned. I implemented to make "Show Answer to a Previous Activity First" feature work with multiple choice, and the multiple choice question cannot remove the last option now.

Frozemint commented:
Thanks, the functionalities I mentioned works now. Would it be possible to realign the multiple choice checkbox/radio buttons in a vertical column like so? This way the style is more consistent with other multiple choice elements the user might encounter from other web pages.

![image](https://user-images.githubusercontent.com/11191061/119870357-218b7b00-bed6-11eb-8642-99b451ed55d8.png)
And perhaps it might be a good idea to align the multiple choice boxes in the center, since that's where the majority of the html elements are... makes the interface easier to see and use. I made a crude approximate sketch in the screenshot below.
![image](https://user-images.githubusercontent.com/11191061/119871017-dde54100-bed6-11eb-8be2-aed5e7ed8a51.png)

In addition, the question author textfields currently have placeholders that can become out of date when the user drag them around. It might be a good idea to not specify ordinality in the placeholder and to instead just for all the placeholders to read "multiple choice option text". This way no matter how many options the user adds/delete/reorders, the placeholder still makes sense.
<p float="left">
  <img src="https://user-images.githubusercontent.com/11191061/119871270-23097300-bed7-11eb-88b1-f04765076329.png" width="200" />
  <img src="https://user-images.githubusercontent.com/11191061/119871314-2ac91780-bed7-11eb-8f95-fe75535cb677.png" width="200" /> 
</p>

harry02031 commented:
Hi @Frozemint, I made changes to the interface to take in account for your comments.

Frozemint commented on templates/vue/src/components/modals/NodeModal/forms/ContentForm/MultipleChoiceForm.vue:
```diff
@@ -0,0 +1,231 @@
+      ],
+      nextChoiceRowId: 4,
+      newChoiceRowTitle: '',
+      choiceRowsRadio: [
```
How come the choiceRowsRadio is an array? I'm not sure if I am understanding the code properly here, but this is the variable that tracks the _single_ selected answer for a radio button right? If a radio button only allows for a single selection, why is an array needed?

harry02031 replied:
no, that is what provides the data for the list of options that the author can put in values for radio, the variable that tracks the single selected answer for a radio button is selectedRadioForm 
<img width="594" alt="Screen Shot 2021-05-27 at 3 10 09 PM" src="https://user-images.githubusercontent.com/30706624/119903604-c8cfd880-befd-11eb-9e34-ae8dcb959272.png">

harry02031 commented:
Updated interface styling

Frozemint commented:
Thanks for all the small adjustments and changes you've made Harry. I'm happy with what we have at this point, and I think any further issues, such as possible code refactoring, and more tweaks and fixes for minor interface issues that are more obscure could work as a future task and PR. Let's have Aidin take a last look and we are done.

harry02031 commented:
Closing this PR as it's superseded by #1057
